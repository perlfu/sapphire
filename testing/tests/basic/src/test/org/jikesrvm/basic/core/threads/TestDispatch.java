/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Common Public License (CPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/cpl1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

class TestDispatch {

  public static void main(String[] args) {

    final int threadCount = 2;
    final TestDispatchWorker[] workers = new TestDispatchWorker[threadCount];
    for (int i = 0; i < threadCount; i++) {
      workers[i] = new TestDispatchWorker("worker " + i);
    }
    for (final TestDispatchWorker worker : workers) {
      worker.start();
    }

    boolean done = false;
    while (!done) {
      Thread.yield();
      done = true;
      for (final TestDispatchWorker worker : workers) {
        done &= worker.completed;
      }
    }
    XThread.say("bye");
    XThread.outputMessages();
  }

  static class TestDispatchWorker extends XThread {

    TestDispatchWorker(String name) {
      super(name);
    }

    void performTask() {
      for (int i = 0; i < 4; ++i) {
        XThread.say("sleeping");
        try { sleep(1000); } catch (InterruptedException e) {}
        XThread.say("running");
        System.gc();
        XThread.say("gc completed");
      }
    }
  }
}


