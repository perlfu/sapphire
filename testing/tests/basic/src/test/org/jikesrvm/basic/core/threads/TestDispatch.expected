Thread 0 Main: bye
Thread 1 worker 0: creating
Thread 1 worker 0: starting
Thread 1 worker 0: run starting
Thread 1 worker 0: sleeping
Thread 1 worker 0: running
Thread 1 worker 0: gc completed
Thread 1 worker 0: sleeping
Thread 1 worker 0: running
Thread 1 worker 0: gc completed
Thread 1 worker 0: sleeping
Thread 1 worker 0: running
Thread 1 worker 0: gc completed
Thread 1 worker 0: sleeping
Thread 1 worker 0: running
Thread 1 worker 0: gc completed
Thread 1 worker 0: run finishing
Thread 2 worker 1: creating
Thread 2 worker 1: starting
Thread 2 worker 1: run starting
Thread 2 worker 1: sleeping
Thread 2 worker 1: running
Thread 2 worker 1: gc completed
Thread 2 worker 1: sleeping
Thread 2 worker 1: running
Thread 2 worker 1: gc completed
Thread 2 worker 1: sleeping
Thread 2 worker 1: running
Thread 2 worker 1: gc completed
Thread 2 worker 1: sleeping
Thread 2 worker 1: running
Thread 2 worker 1: gc completed
Thread 2 worker 1: run finishing
